### ISSUE TITLE
Neo4j not serializable

### ISSUE THREAD

**Issue by PandaGongfu** [neo4j-contrib/neo4j-spark-connector#71]
Hi, @jexp 

I'm trying to run a cypher query within map, it gave a Spark serialization exception. 

val query = "MATCH (n:Stock) where n.StockId = {x} return n.StockId as stock"
val rdd = sc.parallelize(Array(1,2,3,4,5,6),2)
rdd.map(i=> {
    Neo4j(rdd.sparkContext).cypher(query).param("x", s"S$i").batch(25).loadDataFrame.count
})

Below is the error message.
_org.apache.spark.SparkException: Task not serializable
Caused by: java.io.NotSerializableException: org.neo4j.spark.Neo4j
Serialization stack:
	- object not serializable (class: org.neo4j.spark.Neo4j, value: org.neo4j.spark.Neo4j@63f19f0a)_

**Comment by jexp** [comment]
I'm not sure why it tries to serialize the Neo4j Object in this case.
As you're not passing it in from the outside and it is fully contained in the `map()`

But I think we could make Neo4j also serializable, there should be nothing prohibiting that.

**Comment by PandaGongfu** [comment]
@jexp , How do we make Neo4j serializable? Sorry, I'm fairly new to Spark.

**Comment by jexp** [comment]
Can you try this build: https://www.dropbox.com/s/rbjio9xhggm2a0f/neo4j-spark-connector-2.2.1-M5.jar?dl=0

And let me know if it helps.

**Comment by PandaGongfu** [comment]
@jexp,

This is the build that I was already using with Databricks though. Any other ideas?

<img width="300" alt="screen shot 2018-08-28 at 9 57 10 am" src="https://user-images.githubusercontent.com/9537728/44727823-e8b1e380-aaa8-11e8-99ee-029b0cc98a53.png">

**Comment by jexp** [comment]
It's a local build that I rebuilt here, it just has the same version number.

**Comment by PandaGongfu** [comment]
Got it. Tried it with the local build. The error message now says it lacks SparkContext. 

val query = "MATCH (n:Stock) where n.StockId = {x} return n.StockId as stock"
val rdd = sc.parallelize(Array(1,2,3,4,5,6),2)
val dfs = Array()
val tests = rdd.map(i=> {
    val df = Neo4j(rdd.sparkContext).cypher(query).param("x", s"S$i").batch(25).loadDataFrame
    i    
})

tests.take(3).foreach(println)

<img width="760" alt="screen shot 2018-08-28 at 2 50 26 pm" src="https://user-images.githubusercontent.com/9537728/44743905-bec0e700-aad1-11e8-971a-4fe7dead4f3a.png">

**Comment by jexp** [comment]
can you use sparkconfig instead which should be a global variable.
spark context should also be available globally

**Comment by PandaGongfu** [comment]
Do you mean this? It still has the same error though. 

val tests = rdd.map(i=> {
    val df = Neo4j(sc).cypher(query).param("x", s"S$i").batch(25).loadDataFrame
    i    
})

**Comment by PandaGongfu** [comment]
@jexp 

I think I did not attach the local build properly yesterday. After re-attaching the jar file, I got the SparkContext not serializable error. Then I tried to make it serializable, but still got the same error. 

Any comment on the code below?

class Holder extends java.io.Serializable {
  def get_neo_driver() = {
    Neo4j(sc)      
  }
}

val query = "MATCH (n:Stock) where n.StockId = {x} return n.StockId as stock"
val rdd = sc.parallelize(Array(1,2,3,4,5,6),2)
val dfs = Array()

val tests = rdd.map(i=> {
    val driver_class = new Holder()
    val neo = driver_class.get_neo_driver()
    val df = neo.cypher(query).param("x", s"S$i").batch(25).loadDataFrame
    i    
})

tests.take(3).foreach(println)

<img width="703" alt="screen shot 2018-08-29 at 11 23 20 am" src="https://user-images.githubusercontent.com/9537728/44797821-1ddc3580-ab7e-11e8-9a71-b4287b7dc6f5.png">
